### PR TITLE
feat(Paths): Add optional `anyArrayIndexAccessor`

### PR DESCRIPTION
**Pull request by mylesj**
## PR Checklist

- [x] Addresses an existing open issue: related to 
    - resolves: #411
- [x] Steps in [Contributing](https://github.com/ts-essentials/ts-essentials/blob/master/CONTRIBUTING.md) were taken

## Overview

Adds optional configuration that allows wildcards to satisfy array index access.

### CODE REVIEW

File: test/paths.ts [hunk 1]
```diff
@@ -186,3 +186,27 @@ function testDepth2() {
...
+  type Without = Paths<ComplexNestedRequired>;
+
+  type cases_with_any_accessor = [
+    Assert<Has<With, "nested.array.*">>,
+    Assert<Has<With, "nested.array.*.bar">>,
+    Assert<Has<With, `nested.array.${number}`>>,
+    Assert<Has<With, `nested.array.${number}.bar`>>,
+    Assert<Has<With, "nested.tuple.0">>,
+    Assert<Has<With, "nested.tuple.2.good">>,
+  ];
```
Comment (mylesj) [inline test/paths.ts:204]: I felt like adding this might be appropriate as I'm not sure that `nested.tuple.*` makes semantic sense. 

```typescript
    // @ts-expect-error: tuples should not be wildcard-able
    Assert<Has<With, "nested.tuple.*">>,
    // @ts-expect-error: tuples should not be wildcard-able
    Assert<Has<With, "nested.tuple.*.good">>,
```

However I could not get it working, trying several avenues of differentiating tuples from arrays - I must be misunderstanding something. This doesn't block my application but if you agree and would rather conservative exposure, I'm going to need some help with it.

File: tsconfig.test.json [hunk 2]
```diff
@@ -1,4 +1,4 @@
 {
   "extends": "./tsconfig.json",
-  "include": ["lib/**/*.ts"]
-}
+  "include": ["lib/**/*.ts", "test/**/*.ts"]
```
Comment (mylesj) [inline tsconfig.test.json:3]: This is a drive-by fix - the tests are quite neat but they weren't actually running for me. I believe this was the intent?

File: lib/paths/index.ts [hunk 3]
```diff
@@ -20,25 +21,33 @@ type DefaultRecursivePathsOptions = {
  */
 type RecursivePathsOptions = {
   depth: any[];
+  anyArrayIndexAccessor: string | false;
```
Comment (Beraliv) [inline lib/paths/index.ts:24]: `RecursivePathsOptions` is required for `CallOptions`. Since you don't accumulate anything during recursive calls (unlike `depth` where type will contain the number of recursive calls), I suggest removing the value here

File: lib/paths/index.ts [hunk 4]
```diff
@@ -54,7 +63,7 @@ type RecursivePaths<
       ValueOf<{
         [Key in keyof Type]: Key extends Pathable
           ?
-              | `${Key}`
+              | `${AnyArrayIndexAccessorOrKey<Key, CallOptions>}`
```
Comment (Beraliv) [inline lib/paths/index.ts:66]: Since you don't update `anyArrayIndexAccessor` in `CallOptions` (unlike `depth`), I suggest using `UserOptions`:

```suggestion
              | `${AnyArrayIndexAccessorOrKey<Key, UserOptions>}`
```

File: lib/paths/index.ts [hunk 5]
```diff
@@ -67,11 +76,12 @@ type RecursivePaths<
                           UserOptions,
                           {
                             depth: Append<CallOptions["depth"]>;
+                            anyArrayIndexAccessor: AnyArrayIndexAccessor<Value, UserOptions>;
```
Comment (Beraliv) [inline lib/paths/index.ts:79]: As per comment above, `anyArrayIndexAccessor` is static, I suggest removing this line:

```suggestion
```

File: lib/paths/index.ts [hunk 6]
```diff
@@ -94,6 +104,17 @@ type UnsafePaths<Type, Options extends Required<PathsOptions>> = Type extends Ty
     : never
   : never;
 
+type AnyArrayIndexAccessor<Type, Options extends Required<PathsOptions>> = Type extends any[]
```
Comment (Beraliv) [inline lib/paths/index.ts:107]: You can safely remove type `AnyArrayIndexAccessor` since `anyArrayIndexAccessor` is always a string

File: lib/paths/index.ts [hunk 7]
```diff
@@ -94,6 +104,17 @@ type UnsafePaths<Type, Options extends Required<PathsOptions>> = Type extends Ty
...
+  ? Options["anyArrayIndexAccessor"] extends string
+    ? `${Options["anyArrayIndexAccessor"]}`
+    : false
+  : false;
+
+type AnyArrayIndexAccessorOrKey<
+  Key extends Pathable,
+  Options extends RecursivePathsOptions,
+> = Options["anyArrayIndexAccessor"] extends false ? Key : Key | Options["anyArrayIndexAccessor"];
+
```
Comment (Beraliv) [inline lib/paths/index.ts:117]: You can safely remove boolean conditional types since the config option is always a string.

File: test/paths.ts (hunk 1 above)
Comment (Beraliv) [reply test/paths.ts:204]: Give `` `${number}` `` a try as I recommended above, it should help.

Tuples have specific keys: they are numeric literal types (e.g. 0, 1, 2). A conditional type with `DefaultPathsOptions["anyArrayIndexAccessor"]` and `` `${Key}` `` in `type AnyArrayIndexAccessorOrKey` should be sufficient to differentiate tuples and arrays

File: tsconfig.test.json (hunk 2 above)
Comment (Beraliv) [reply tsconfig.test.json:3]: Good spot, I broke it a couple of commits ago! Thanks for fixing it!

File: test/paths.ts [hunk 8]
```diff
@@ -186,3 +186,27 @@ function testDepth2() {
...
+ * Configuration allows wildcard accessors for array indices.
+ */
+function testAnyArrayIndexAccessor() {
+  type With = Paths<ComplexNestedRequired, { anyArrayIndexAccessor: "*" }>;
+  type Without = Paths<ComplexNestedRequired>;
+
+  type cases_with_any_accessor = [
+    Assert<Has<With, "nested.array.*">>,
+    Assert<Has<With, "nested.array.*.bar">>,
+    Assert<Has<With, `nested.array.${number}`>>,
```
Comment (Beraliv) [inline test/paths.ts:200]: During my testing, I found that `"nested.array.*"` will be acceptable, but not `` `nested.array.${number}` ``. I couldn't figure it out why this happens, so if you face it during your testing and cannot figure it out how to fix it, please leave it. It's a known bug

**Review by Beraliv — CHANGES_REQUESTED** [review]
@mylesj Hey! Thank you so much for taking your time and raising this PR! Your PR looks good, I suggested multiple changes to align with my expectations. Let me know what you think

**Comment by Beraliv** [comment]
@mylesj for your benefit, the testing strategy of ts-essentials is to run typechecks on multiple versions (between 4.5 and 5.5). Currently, 3 typechecks have failed on version 4.6-4.8, as GHA doesn't contain `continue-on-error: true`, it will stop running others. Kindly ping me when you're ready, I will accept re-running typechecks once again.

File: test/paths.ts (hunk 1 above)
Comment (mylesj) [reply test/paths.ts]: Ah of course, this makes more sense.

I was trying a combination along the lines of this and it would either always swallow all arrays or all tuples:
```typescript
IsTuple<Type> extends Type 
  ? A 
  : Type extends any[]
  ? B
  : C
```

But I agree, the condition ends up much simpler when the `Key` is already inferred as `number`.

File: test/paths.ts (hunk 8 above)
Comment (mylesj) [reply test/paths.ts:200]: Can you confirm if this is still an issue following the update and if so which version of TypeScript?

**Comment by mylesj** [comment]
> _@mylesj for your benefit, the testing strategy of ts-essentials is to run typechecks on multiple versions (between 4.5 and 5.5). Currently, 3 typechecks have failed on version 4.6-4.8, as GHA doesn't contain continue-on-error: true, it will stop running others. Kindly ping me when you're ready, I will accept re-running typechecks once again._

@Beraliv So I had a look at the tests under the different versions and I'm reasonably sure this change set isn't directly related - more a consequence of my re-enabling the tests and surfacing existing errors on `master`. Specifically: 
  - `test/paths.ts` fail on **4.6**
  - `test/xor.ts` fail on: **4.6**, **4.7** & **4.8**

That said, I spent some time trying to understand the paths issue but can't for the moment see what the problem might be - something is truncating the paths prematurely or preventing recursion. I'd ask that you take a closer look to confirm whether this is related to recent changes.

If you agree, I've selectively disabled those specific test suites for now. Otherwise I'm going to be unavailable for the next week or so but will take another look when I'm back.

File: test/paths.ts (hunk 8 above)
Comment (Beraliv) [reply test/paths.ts:200]: I'll double-check it on the weekend

File: lib/paths/index.ts [hunk 9]
```diff
@@ -78,7 +80,9 @@ type RecursivePaths<
                         > extends infer Rest
                         ? IsNever<Rest> extends true
                           ? never
-                          : `${AnyArrayIndexAccessorOrKey<Key, UserOptions>}.${Rest & string}`
+                          : Rest extends ReallyPrimitive
+                          ? `${AnyArrayIndexAccessorOrKey<Key, UserOptions>}.${Rest}`
+                          : never
```
Comment (mylesj) [inline lib/paths/index.ts:85]: @Beraliv 

I think I got it, this `Rest & string` is broadening the path to any `anything.nested.${string}` in earlier versions.

File: test/paths.ts (hunk 8 above)
Comment (Beraliv) [reply test/paths.ts:200]: When I enabled tests in another branch (which is created from `master` branch), I could reproduce the problem: commit https://github.com/ts-essentials/ts-essentials/pull/413/commits/64fffdc70bf9a5e86482fadb894c8604e7baa92b

Screenshot 2024-10-07 at 09 31 25

The problem existed in TypeScript versions 4.5-4.8, it's fixed after this commit - https://github.com/ts-essentials/ts-essentials/pull/413/commits/98258d5fbccd4e6d0ba4d3e32ec58ce42c9f33ab

I'll get it merged today, please rebase it and you won't need to make any other changes

File: lib/paths/index.ts [hunk 10]
```diff
@@ -69,9 +77,9 @@ type RecursivePaths<
                             depth: Append<CallOptions["depth"]>;
                           }
                         > extends infer Rest
-                        ? IsNever<Rest> extends true
-                          ? never
-                          : `${Key}.${Rest & string}`
+                        ? Rest extends Exclude<Primitive, symbol>
```
Comment (Beraliv) [inline lib/paths/index.ts:80]: You may look for `Pathable` here:

```suggestion
                        ? Rest extends Pathable
```

I've tested it in https://github.com/ts-essentials/ts-essentials/pull/413 (commit https://github.com/ts-essentials/ts-essentials/pull/413/commits/98258d5fbccd4e6d0ba4d3e32ec58ce42c9f33ab) and it fixed the problem for TypeScript versions 4.5-4.8

File: tsconfig.test.json [hunk 11]
```diff
@@ -1,4 +1,8 @@
 {
   "extends": "./tsconfig.json",
-  "include": ["lib/**/*.ts"]
-}
+  "include": ["lib/**/*.ts", "test/**/*.ts"],
+  "exclude": [
+    // TODO: fix 4.6.4; 4.7.4; 4.8.4
+    "test/xor.ts",
+  ]
```
Comment (Beraliv) [inline tsconfig.test.json:7]: Please remove it when https://github.com/ts-essentials/ts-essentials/pull/413 is merged

**Review by Beraliv — COMMENTED** [review]
3 more comments, otherwise LGTM!

**Comment by mylesj** [comment]
Fantastic. So I'm AFK / on vacation this week but will wrap things up at the weekend.

File: tsconfig.test.json (hunk 11 above)
Comment (Beraliv) [reply tsconfig.test.json:7]: The PR is just merged ✅

File: lib/paths/index.ts [hunk 12]
```diff
@@ -96,6 +103,12 @@ type UnsafePaths<Type, Options extends Required<PathsOptions>> = Type extends Ty
     : never
   : never;
 
+// prettier-ignore
```
Comment (Beraliv) [inline lib/paths/index.ts:106]: Is it still required?

Comment (mylesj) [reply lib/paths/index.ts:106]: No, I just thought it read better - the line length seems to be quite long. I'll remove it though.

**Review by Beraliv — APPROVED** [review]
Great work! Thank you for your efforts, LGTM now 👏

**Review by Beraliv — COMMENTED** [review]
Last thing: please add a changeset (part of contributing guide): you have to run `yarn changeset` and add a minor change with your preferred message

**Comment by mylesj** [comment]
Ah my bad, I'm not familiar with this approach to change log. Done - thanks for sticking with me, appreciate the support!

**Comment by Beraliv** [comment]
Great, thank you @mylesj for your efforts! Changes look good, so I will merge it now
